azurelinuxagent/ga/remoteaccess.py

```diff
@@ -0,0 +1,163 @@
+                self.protocol.client.update_goal_state(True)
+                self.protocol.client.update_remote_access_conf(self.protocol.client.goal_state)
+                if self.remote_access is None or self.remote_access.incarnation != self.protocol.client.remote_access.incarnation:
+                    self.remote_access = self.protocol.client.remote_access
+                    self.handle_remote_access()
+            else:
+                logger.info("Non JIT enabled client.")
+        except Exception as e:
+            msg = u"Exception processing remote access handler: {0}".format(
+                ustr(e))
+            logger.warn(msg)
+            logger.info(str(e))
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.HandleRemoteAccess,
+                      is_success=False,
+                      message=msg)
+            return     
+
+    def handle_remote_access(self):
+        logger.verbose("Entered handle_remote_access")        
+        if self.remote_access is not None:
+            # Get JIT user accounts.
+            all_users = self.os_util.get_users()
+            jit_users = set()
+            for usr in all_users:
+                if usr[4] == REMOTE_ACCESS_ACCOUNT_COMMENT:
+                    jit_users.add(usr[0])     
+            for acc in self.remote_access.user_list.users:
+                raw_expiration = acc.expiration
+                account_expiration = datetime.strptime(raw_expiration, REMOTE_USR_EXPIRATION_FORMAT)
+                now = datetime.utcnow()
+                if acc.name not in jit_users and now < account_expiration:
+                    self.add_user(acc.name, acc.encrypted_password, account_expiration)
+
+    def add_user(self, username, encrypted_password, account_expiration, retry=MAX_TRY_ATTEMPT, throttle=FAILED_ATTEMPT_THROTTLE):
+        created = False
+        attempt_count = 0
+        while attempt_count < retry:
+            try:
+                attempt_count += 1
+                expiration_date = account_expiration + timedelta(days=1) - datetime.utcnow()
+                logger.verbose("add_user: Adding user {0} with expiration in {1} day(s)".format(username, expiration_date.days))
+                self.os_util.useradd(username, expiration_date.days, REMOTE_ACCESS_ACCOUNT_COMMENT)
+                created = True
+                cache_file = os.path.join(conf.get_lib_dir(), "remote_access_pwd.dat")
+                prv_key = os.path.join(conf.get_lib_dir(), TRANSPORT_PRIVATE_CERT)
+                pwd = self.cryptUtil.decrypt_secret(encrypted_password, prv_key, cache_file, None)
+                self.os_util.chpasswd(username, pwd, conf.get_password_cryptid(), conf.get_password_crypt_salt_len())
+                self.os_util.conf_sudoer(username)
+                logger.info("add_user: User '{0}' added successfully with expiration in {1} day(s)".format(username, expiration_date.days))
```

Comment:
>logger.info("add_user: User '{ [](start = 16, length = 30)

It may be valuable to put the expiration time.  The user can construct it from the log timestamp, but there's no need to make people think. :)

Reply:
Until I implement delete, the expiration date in JIT will not match the expiration that will actually happen on the box since it only allows a date or number of days,   Often JIT accounts are only provisioned for a few hours, while we won't expire them here until the following day.

---
In reply to: [192216983](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192216983) [](ancestors = 192216983)